### ISSUE TITLE
Stopping Windows service doesn't stop servers

### ISSUE THREAD

**Issue by bcat** [PufferPanel/PufferPanel#1149]
I was looking for a game server manager for some lightweight home use (hosting a couple Minecraft and Terraria servers for friends), and I found PufferPanel. I was super excited that it had a native Windows service, and indeed it seems to work wonderfully! However, I have noticed one issue that can potentially lead to server data loss.

Specifically, when the Windows service is stopped (or when the service is restarted, or when the system is shut down), any running game servers continue to run. This means that game data won't be saved, and if the service is restarted, there will be orphaned server instances in the background it's not tracking.

As far as I can tell, `runservice.go` tells PufferPanel to terminate when the service is stopped, but PufferPanel doesn't take any action to shut down running servers. IMO it should do so, or at least have an option to do so.

I'm not familiar with the codebase, but is [`programs.ShutdownService`](https://github.com/PufferPanel/PufferPanel/blob/c7a02698500b6a840f4761eb283d58ad4d907f65/programs/program.go#L68) the right place to do this?

Interestingly, when I run PufferPanel in the Windows console via `run.go` and send Ctrl+C to the console, running game servers _do_ terminate cleanly (as confirmed in their log files, as well as by the lack of orphaned game server processes in that case). I believe that's because [Windows sends Ctrl+C to all processes attached to the console](https://docs.microsoft.com/en-us/windows/console/ctrl-c-and-ctrl-break-signals), and Minecraft (as well as Terraria with the TShock mod) handler Ctrl+C by saving & exiting gracefully.

That means I should actually be able to work around this bug for now by running the non-service variant (`--run` rather than `--runService`) via [NSSM](https://nssm.cc/), which sends Ctrl-C to the underlying application's Windows console when the service is stopped. However, that's a bit silly to do when there's a native Windows service that could be used instead. :)

**Comment by bcat** [comment]
Although I think my "run in NSSM" workaround only helps for game servers that handle Ctrl+C properly (e.g., by saving data). For example, vanilla Terraria does not, so it would still exit abruptly (without saving).

In any case, issuing each program's specific "stop" action seems safer (especially on Windows where, to my knowledge, `SIGTERM` isn't quite a thing).

**Comment by nepcore** [comment]
PufferPanel does actually handle Ctrl+C itself and shuts down running servers before quitting, usually at least, it not happening with the Windows service sounds to me like a bug/oversight

The entire `programs` package is unrelated to PufferPanels own lifecycle, all code in there runs, stops, etc servers on the node  
I'd say `cmd/runservice.go` is where you want to look for the Windows service shutdown, though I may be wrong

**Comment by bcat** [comment]
Hmm yeah, I very likely just missed something in the code. (Just gave a quick skim, haven't looked in detail yet.)

I found what appears to be the [SIGINT handler](https://github.com/PufferPanel/PufferPanel/blob/c7a02698500b6a840f4761eb283d58ad4d907f65/cmd/run.go#L70-L79), and that _does_ work fine on Windows. If you send Ctrl-C to the console the app is attached to, you get the "Shutting down..." message and PufferPanel terminates cleanly.

FWIW, it looks like the service [indicates on the same channel](https://github.com/PufferPanel/PufferPanel/blob/c7a02698500b6a840f4761eb283d58ad4d907f65/cmd/runservice.go#L48) for PufferPanel to terminate when the service stops. And the service does stop in a timely fashion (e.g., it doesn't time out or anything when I stop it in the Windows Services panel). It's just that the game servers keep running in the background in this case.

I couldn't find any code that actually cleans up running game servers when PufferPanel terminates. The only termination code I could find was [this](https://github.com/PufferPanel/PufferPanel/blob/c7a02698500b6a840f4761eb283d58ad4d907f65/cmd/run.go#L128-L143) (which is what made me look into `programs.ShutdownService`). But I'm probably missing something; it's a big codebase and I've only looked at a tiny bit so far. :)

**Comment by bcat** [comment]
So I dug into this a little bit more, and I still can't find anything in PufferPanel itself that explicitly stops running servers when the `pufferpanel` process terminates. However, sometimes the severs terminate for other reasons. I tried these cases:

1. (Tested) Linux, running PufferPanel in a terminal, hitting Ctrl-C: PufferPanel exits, but servers remain running in the background. This makes sense since Linux sends SIGINT to the [foreground process group of the terminal](https://unix.stackexchange.com/a/149756) on Ctrl-C, but it looks like on Linux, PufferPanel [attaches servers to their own pseudoterminal](https://github.com/PufferPanel/PufferPanel/blob/e56451a498443ed93f4d7adade40a509d0640a04/environments/tty/tty.go#L74), so they don't receive the Ctrl-C signal.

2. (Tested) Windows, running PufferPanel in a console window, hitting Ctrl-C: PufferPanel exits, and servers exit too. Also, if the servers handle Ctrl-C, then can terminate gracefully (saving, etc.). This behavior is because [Windows triggers Ctrl-C handlers in all processes attached to that particular console](https://docs.microsoft.com/en-us/windows/console/ctrl-c-and-ctrl-break-signals), similar to how Linux sends SIGINT to the foreground process group. The difference is that on Windows, PufferPanel doesn't create a new console for the server processes by [passing the `CREATE_NEW_CONSOLE` flag when then processes are created](https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags).

3. (Tested) Windows, running PufferPanel as a service, stopping the service: PufferPanel exits, but servers remain running in the background. As far as I know, Windows has no special handling of background processes started by a Windows service, so without PufferPanel doing something explicit, they just keep running.

4. (Not yet tested) Linux, running PufferPanel as a systemd service, stopping the service: I haven't tried this yet, but I expect that PufferPanel will exit (thanks to the service's [`ExecStop` config](https://github.com/PufferPanel/PufferPanel/blob/e56451a498443ed93f4d7adade40a509d0640a04/systemd/servicefiles/pufferpanel.service#L8)), and then the background servers will be sent `SIGTERM` (and exit gracefully if they handle that signal) because the default systemd [`KillMode=control-group`](https://www.freedesktop.org/software/systemd/man/systemd.kill.html) setting causes systemd to send `SIGTERM` to all processes spawned by the service.

So in summary, on Windows, background servers get stopped due to Ctrl-C in console, but not when running as a service. On Linux, the exact opposite is the case, with background servers getting stopped when running as a service, but not when receiving Ctrl-C in the console.

IMO, it would make sense for PufferPanel itself to take care of stopping servers running servers when it exits to ensure consistent behavior no matter how it's run. What do y'all think?

**Comment by bcat** [comment]
If PufferPanel is changed to stop running servers when it terminates, then I think it it would also make sense to pass `CREATE_NEW_PROCESS_GROUP` in [`SysProcAttr.CreationFlags`](https://pkg.go.dev/syscall?GOOS=windows#SysProcAttr) when creating the server processes on Windows to prevent them from receiving Ctrl-C from `pufferpanel`'s console.  This would, if I understand correctly, make the "running in console window" behavior consistent between Windows and Linux. (Otherwise, there may be a race condition between PufferPanel trying to stop the servers itself and the servers responding to the Ctrl-C signal.)

**Comment by LordRalex** [comment]
I feel like not stopping servers was missed, and it just "happened" to work before. That was something I expected we did, but I don't know how it was missed.

What I think I want to do is just have the shutdown call (https://github.com/PufferPanel/PufferPanel/blob/c7a02698500b6a840f4761eb283d58ad4d907f65/programs/program.go#L68) just do a full shutdown of servers, so that they shut down regardless of the OS.

**Comment by bcat** [comment]
Confirmed that servers are now shut down gracefully when stopping the service. Thanks for the fix!!
